wgpu-core/src/command/clear.rs

```diff
@@ -228,68 +221,126 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
             });
         }
 
-        // query from tracker with usage (and check usage)
-        let (dst_texture, dst_pending) = cmd_buf
-            .trackers
-            .textures
-            .use_replace(
-                &*texture_guard,
-                dst,
-                TextureSelector {
-                    levels: subresource_range.base_mip_level..subresource_level_end,
-                    layers: subresource_range.base_array_layer..subresource_layer_end,
-                },
-                hal::TextureUses::COPY_DST,
-            )
-            .map_err(ClearError::InvalidTexture)?;
-        let dst_raw = dst_texture
-            .inner
-            .as_raw()
-            .ok_or(ClearError::InvalidTexture(dst))?;
-        if !dst_texture.desc.usage.contains(TextureUsages::COPY_DST) {
-            return Err(ClearError::MissingCopyDstUsageFlag(None, Some(dst)));
+        clear_texture(
+            &Stored {
+                value: id::Valid(dst),
+                ref_count: dst_texture.life_guard().ref_count.as_ref().unwrap().clone(),
+            },
+            dst_texture,
+            TextureInitRange {
+                mip_range: subresource_range.base_mip_level..subresource_level_end,
+                layer_range: subresource_range.base_array_layer..subresource_layer_end,
+            },
+            cmd_buf.encoder.open(),
+            &mut cmd_buf.trackers.textures,
+            &device_guard[cmd_buf.device_id.value],
+        )
+    }
+}
+
+pub(crate) fn clear_texture<A: hal::Api>(
+    dst_texture_id: &Stored<TextureId>,
+    dst_texture: &Texture<A>,
+    range: TextureInitRange,
+    encoder: &mut A::CommandEncoder,
+    texture_tracker: &mut ResourceTracker<TextureState>,
+    device: &Device<A>,
+) -> Result<(), ClearError> {
+    clear_texture_no_device(
+        dst_texture_id,
+        dst_texture,
+        range,
+        encoder,
+        texture_tracker,
+        &device.alignments,
+        &device.zero_buffer,
+    )
+}
+
+pub(crate) fn clear_texture_no_device<A: hal::Api>(
+    dst_texture_id: &Stored<TextureId>,
+    dst_texture: &Texture<A>,
+    range: TextureInitRange,
+    encoder: &mut A::CommandEncoder,
+    texture_tracker: &mut ResourceTracker<TextureState>,
+    alignments: &hal::Alignments,
+    zero_buffer: &A::Buffer,
+) -> Result<(), ClearError> {
+    let dst_raw = dst_texture
+        .inner
+        .as_raw()
+        .ok_or(ClearError::InvalidTexture(dst_texture_id.value.0))?;
+
+    // Issue the right barrier.
+    let clear_usage = match dst_texture.clear_mode {
+        TextureClearMode::BufferCopy => hal::TextureUses::COPY_DST,
+        TextureClearMode::RenderPass(_) => {
+            if dst_texture
```

Comment:
this may be a little be tricky in the future. Imagine a case where the texture is created in such a way that it has a depth format but can be reinterpret into R32F and rendered into. In this case, the texture itself may have both `COLOR_TARGET` and `DEPTH_STENCIL_WRITE` usages. We don't necessarily need to address this right here, since we don't know if the time will come for this to be possible.

Would it make sense to put that usage bit into `TextureClearMode::RenderPass`? Like `is_color: bool` or something like that.

Reply:
I do like the idea of putting it on the clear_mode, makes this piece and one below more concise; going with that right away